server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -149,34 +148,32 @@ protected WritePrimaryResult<ShardUpsertRequest, ShardResponse> processRequestIt
         ShardResponse shardResponse = new ShardResponse(request.returnValues());
         String indexName = request.index();
         DocTableInfo tableInfo = schemas.getTableInfo(RelationName.fromIndexName(indexName), Operation.INSERT);
-        Reference[] insertColumns = request.insertColumns();
         var mapperService = indexShard.mapperService();
         Function<ColumnIdent, FieldType> getFieldType = column -> mapperService.getLuceneFieldType(column.fqn());
         TransactionContext txnCtx = TransactionContext.of(request.sessionSettings());
-        Indexer indexer = insertColumns == null
+        UpdateToInsert updateToInsert = request.updateColumns() == null
             ? null
-            : new Indexer(
-                indexName,
-                tableInfo,
-                txnCtx,
-                nodeCtx,
-                getFieldType,
-                List.of(insertColumns),
-                request.returnValues()
-            );
-
-        InsertSourceGen insertSourceGen = insertColumns != null && insertColumns.length == 1 && insertColumns[0].column().equals(DocSysColumns.RAW)
-            ? InsertSourceGen.of(txnCtx, nodeCtx, tableInfo, indexName, request.validation(), Arrays.asList(insertColumns))
+            : new UpdateToInsert(nodeCtx, txnCtx, tableInfo, request.updateColumns());
+        List<Reference> insertColumns;
+        if (request.insertColumns() == null) {
+            insertColumns = updateToInsert.columns();
```

Comment:
double checking, insertColumns & updateColumns cannot be both null right?

Reply:
Only if there is a programming error in the plan.execute

Reply:
I think if so it would have to go into the request class to raise the error early where it happens. But having an NPE here was good enough to easily spot the mistake. I had that happen while working on it.